### ISSUE TITLE
Pydantic Dataclass not fully supported

### ISSUE THREAD

**Issue by justinaustin** [microsoft/pyright#1510]
**Is your feature request related to a problem? Please describe.**
Currently pyright treats all classes wrapped with the [`@pydantic.dataclasses.dataclass` decorator](https://pydantic-docs.helpmanual.io/usage/dataclasses/) as of the same type `Dataclass`, instead of reading the declared attributes and types. Whereas for the standard library dataclasses, pyright correctly recognizes their type and attribute types.

```python
from dataclasses import dataclass
from pydantic.dataclasses import dataclass as pydantic_dataclass

@dataclass
class Foo:
    x: int

@pydantic_dataclass
class Bar:
    x: int

foo = Foo(1)
bar = Bar(1)

foo.x  # type checks correctly
bar.x  # errors with `Cannot access member "x" for type "Dataclass"`
```

**Describe the solution you'd like**
Ideally, pyright would treat pydantic dataclasses the same (or nearly the same) way as it treats dataclasses from the standard library. `mypy` has this same problem until you install [pydantic's mypy plugin](https://pydantic-docs.helpmanual.io/mypy_plugin/), and then it works as expected. It doesn't [look](https://github.com/microsoft/pyright/issues/947#issuecomment-674534268) like pyright has a plugin architecture yet, but would there still be a way to account for this behavior?

Thank you!

**Comment by erictraut** [comment]
Pyright doesn't contain specialized knowledge of any third-party libraries. That would be a slippery slope. This is a hard rule that we're not willing to violate.

We've talked about possibly adding support for plugins at some point, but we'd want to do so in a way that they would work with other type checkers as well. It doesn't make sense for library authors (or the community) to maintain different plugins for different type checkers. But creating a standard for type checker plugins (or put another way, extensions to the Python type system) is a heavy lift, and I don't see it happening any time soon.

My recommendation is to avoid using third-party libraries that implement non-standard behaviors, i.e. behaviors that cannot be described by existing Python type annotations. In the particular case of pydantic, I recommend switching to the stdlib `dataclass` if you can. I realize that may not be feasible for existing code bases, but I think it's good advice for new code.

**Comment by erictraut** [comment]
Pyright doesn't contain specialized knowledge of any third-party libraries. That would be a slippery slope. This is a hard rule that we're not willing to violate.

We've talked about possibly adding support for plugins at some point, but we'd want to do so in a way that they would work with other type checkers as well. It doesn't make sense for library authors (or the community) to maintain different plugins for different type checkers. But creating a standard for type checker plugins (or put another way, extensions to the Python type system) is a heavy lift, and I don't see it happening any time soon.

My recommendation is to avoid using third-party libraries that implement non-standard behaviors, i.e. behaviors that cannot be described by existing Python type annotations. In the particular case of pydantic, I recommend switching to the stdlib `dataclass` if you can. I realize that may not be feasible for existing code bases, but I think it's good advice for new code.

**Comment by justinaustin** [comment]
Thank you for the quick response. Our codebase currently gets a lot of value out of pydantic's automatic runtime checking of the dataclass types, so I don't see us using the stdlib `dataclass` anytime soon. But I understand pyright not wanting to have special knowledge of third party libraries.

Thanks again!

**Comment by layday** [comment]
You could probably get away with `if TYPE_CHECKING: pydantic_dataclass = dataclass` here.  The public interface of the built-in dataclass and the Pydantic dataclass is the same; what Pydantic does is customise `__post_init__` and add a bunch of private attributes of its own to the class.

**Comment by vpeopleonatank** [comment]
> You could probably get away with `if TYPE_CHECKING: pydantic_dataclass = dataclass` here. The public interface of the built-in dataclass and the Pydantic dataclass is the same; what Pydantic does is customise `__post_init__` and add a bunch of private attributes of its own to the class.

I can't find `pydantic_dataclass`. Can you add more detail?

**Comment by layday** [comment]
```py
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    pydantic_dataclass = dataclass
else:
    from pydantic.dataclasses import dataclass as pydantic_dataclass

@pydantic_dataclass
class Foo:
    ...
```
